File: NotificationMS/src/main/java/com/alumniconnect/notification/controller/NotificationController.java
package com.alumniconnect.notification.controller;

import java.util.List;

import org.springframework.web.bind.annotation.*;

import com.alumniconnect.notification.dto.NotificationDto;
import com.alumniconnect.notification.entity.Notification;
import com.alumniconnect.notification.service.NotificationService;

import jakarta.validation.Valid;

@RestController
@RequestMapping("/notifications")
public class NotificationController {

    private final NotificationService notificationService;

    public NotificationController(NotificationService notificationService) {
        this.notificationService = notificationService;
    }

    // Create Notification
    @PostMapping
    public NotificationDto createNotification(
            @Valid @RequestBody NotificationDto dto) {

        return notificationService.createNotification(dto);
    }

    // Get All Notifications
    @GetMapping
    public List<Notification> getAllNotifications() {

        return notificationService.getAllNotifications();
    }

    // Get Notification By Id
    @GetMapping("/{id}")
    public Notification getNotificationById(
            @PathVariable Long id) {

        return notificationService.getNotificationById(id);
    }

    // Get Notifications of User
    @GetMapping("/user/{userId}")
    public List<Notification> getUserNotifications(
            @PathVariable Long userId) {

        return notificationService.getUserNotifications(userId);
    }

    // Get Read/Unread Notifications
    @GetMapping("/status/{isRead}")
    public List<Notification> getNotificationsByReadStatus(
            @PathVariable Boolean isRead) {

        return notificationService.getNotificationsByReadStatus(isRead);
    }

    // Mark Notification As Read
    @PutMapping("/{id}/read")
    public Notification markAsRead(
            @PathVariable Long id) {

        return notificationService.markAsRead(id);
    }

    // Delete Notification
    @DeleteMapping("/{id}")
    public void deleteNotification(
            @PathVariable Long id) {

        notificationService.deleteNotification(id);
    }
}

File: NotificationMS/Controllers/NotificationController.cs
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using NotificationMS.Data;
using NotificationMS.Models;

namespace NotificationMS.Controllers
{
    [ApiController]
    [Route("api/v1/notifications")]
    public class NotificationController : ControllerBase
    {
        private readonly NotificationDbContext _context;

        public NotificationController(NotificationDbContext context)
        {
            _context = context;
        }

        // Create Notification
        [HttpPost]
        public async Task<ActionResult<Notification>> CreateNotification([FromBody] Notification notification)
        {
            if (notification == null)
            {
                return BadRequest("Notification request body is null.");
            }

            _context.Notifications.Add(notification);
            await _context.SaveChangesAsync();

            return Ok(notification);
        }

        // Get All Notifications
        [HttpGet]
        public async Task<ActionResult<IEnumerable<Notification>>> GetAllNotifications()
        {
            var list = await _context.Notifications.ToListAsync();
            return Ok(list);
        }

        // Get Notification By Id
        [HttpGet("{id}")]
        public async Task<ActionResult<Notification>> GetNotificationById(long id)
        {
            var notification = await _context.Notifications.FindAsync(id);
            if (notification == null)
            {
                return NotFound($"Notification with ID {id} not found.");
            }
            return Ok(notification);
        }

        // Get Notifications of User
        [HttpGet("user/{userId}")]
        public async Task<ActionResult<IEnumerable<Notification>>> GetUserNotifications(long userId)
        {
            var userNotifications = await _context.Notifications
                .Where(n => n.UserId == userId)
                .ToListAsync();
            return Ok(userNotifications);
        }

        // Get Read/Unread Notifications
        [HttpGet("status/{isRead}")]
        public async Task<ActionResult<IEnumerable<Notification>>> GetNotificationsByReadStatus(bool isRead)
        {
            var notifications = await _context.Notifications
                .Where(n => n.IsRead == isRead)
                .ToListAsync();
            return Ok(notifications);
        }

        // Mark Notification As Read
        [HttpPut("{id}/read")]
        public async Task<ActionResult<Notification>> MarkAsRead(long id)
        {
            var notification = await _context.Notifications.FindAsync(id);
            if (notification == null)
            {
                return NotFound($"Notification with ID {id} not found.");
            }

            notification.IsRead = true;
            await _context.SaveChangesAsync();

            return Ok(notification);
        }

        // Delete Notification
        [HttpDelete("{id}")]
        public async Task<IActionResult> DeleteNotification(long id)
        {
            var notification = await _context.Notifications.FindAsync(id);
            if (notification == null)
            {
                return NotFound($"Notification with ID {id} not found.");
            }

            _context.Notifications.Remove(notification);
            await _context.SaveChangesAsync();

            return Ok();
        }
    }
}
